src/storage/src/source/reclock.rs

```diff
@@ -8,413 +8,668 @@
+            }
+        }
 
-            snapshot_map
-                .into_iter()
-                .map(|(partition, (ts, offset))| {
-                    (
-                        partition,
-                        VecDeque::from_iter(std::iter::once((ts, offset))),
-                    )
-                })
-                .collect()
-        } else {
-            HashMap::new()
-        };
-        let read_cursors = HashMap::new();
+        // Ensure we have enough bindings and re-sync state if we lost a race
+        while let Err(Upper(global_upper)) = self.mint(&batch_upper).await {
+            self.sync(&global_upper).await;
+        }
 
-        let timestamp_bindings_listener = read_handle
-            .listen(as_of)
-            .await
-            .expect("Initial listen at handle.since ts")
-            .into_stream()
-            .boxed();
-
-        Ok(Self {
-            name,
-            read_progress,
-            write_upper,
-            persisted_timestamp_bindings,
-            read_cursors,
-            write_handle,
-            timestamp_bindings_listener,
-            now,
-            update_interval,
-            last_timestamp: Timestamp::minimum(),
-        })
+        ReclockIter {
+            reclock: self,
+            messages: batch.iter_mut(),
+        }
     }
 
-    // TODO: require NowFn to be monotonic so we can simplify this function
-    fn get_time(&mut self) -> Option<Timestamp> {
-        let update_interval: u64 = self
-            .update_interval
-            .clone()
-            .as_millis()
-            .try_into()
-            .expect("interval millis don't fit into u64");
-        let new_ts = (self.now)();
-
-        // Mod and then round up
-        let remainder = new_ts % update_interval;
-        let new_ts_rounded = if remainder == 0 {
-            new_ts
-        } else {
-            new_ts - remainder + update_interval
-        };
+    /// Calling this method promises that all future timestamping requests will be for offsets
+    /// greater or equal to the provided source upper frontier. In exchange the reclock operator
+    /// return the timestamp frontier that the caller can use to downgrade its capability to.
+    pub async fn advance(
+        &mut self,
+        source_upper: &HashMap<PartitionId, MzOffset>,
+    ) -> Antichain<Timestamp> {
+        // First ensure that the bindings cover the provided frontier
+        while let Err(Upper(global_upper)) = self.mint(&source_upper).await {
+            self.sync(&global_upper).await;
+        }
 
-        if new_ts_rounded > self.last_timestamp {
-            self.last_timestamp = new_ts_rounded;
-            Some(new_ts_rounded)
-        } else {
-            None
+        // Then, figure out the highest timestamp the caller we downgrade to, which is the minimum
+        // timestamp of all the bindings involved in the given frontier.
+        let mut handle_since = self.upper.clone();
```

Comment:
I think there might be something implicit in the "what the caller can downgrade capability to" because it's not clear to me that `handle_since` is a valid return value of the offset in `source_upper` is beyond `self.last_source_upper` for all partitions.  Can you add a little more detail to the doc-comment on this function relating the return value to the timestamps that can be returned by `Self::relock` in the future?